### PR TITLE
Stop removal until node is removed from applied spec

### PR DESCRIPTION
**Pull request by galal-hussein**
https://github.com/rancher/rancher/issues/11093

### CODE REVIEW

**Comment by galal-hussein** [comment]
When deleting any machine, it will be removed almost immediately before rke gets the chance to clean the node, this will create a problem with etcd, which will cause the etcd to fail and not removed gracefully, this PR will stop removal of the node until its removed from the applied spec.

File: controller/machine/controller.go [hunk 1]
```diff
@@ -53,6 +54,7 @@ type Lifecycle struct {
 	machineTemplateClient        v3.MachineTemplateInterface
 	configMapGetter              typedv1.ConfigMapsGetter
 	logger                       event.Logger
+	clusters                     v3.ClusterInterface
```
Comment (alena1108) [inline controller/machine/controller.go]: we should use cache here (the lister). Unless there is a need to pull from api directly

File: controller/machine/controller.go [hunk 2]
```diff
@@ -300,3 +310,24 @@ func (m *Lifecycle) saveConfig(config *machineconfig.MachineConfig, machineDir s
...
+
+func (m *Lifecycle) isNodeInAppliedSpec(machine *v3.Machine) (bool, error) {
+	cluster, err := m.clusterLister.Get("", machine.Spec.ClusterName)
+	if err != nil {
+		return false, err
+	}
+	if cluster == nil {
+		return false, nil
+	}
+	if cluster.Status.AppliedSpec.RancherKubernetesEngineConfig == nil {
```
Comment (alena1108) [inline controller/machine/controller.go]: lets also check that the cluster.deletiontimestamp is nil. If not, return false right away

File: controller/machine/controller.go [hunk 3]
```diff
@@ -300,3 +310,24 @@ func (m *Lifecycle) saveConfig(config *machineconfig.MachineConfig, machineDir s
...
+		return false, nil
+	}
+
+	for _, node := range cluster.Status.AppliedSpec.RancherKubernetesEngineConfig.Nodes {
+		machineName := strings.Split(node.MachineName, ":")
+		if machineName[1] == machine.Name {
```
Comment (alena1108) [inline controller/machine/controller.go]: can machinename ever be empty? lets guard against it, and if len(splitted) < 2, continue

**Comment by alena1108** [comment]
LGTM

File: controller/machine/controller.go [hunk 4]
```diff
@@ -135,7 +137,20 @@ func (m *Lifecycle) Remove(obj *v3.Machine) (*v3.Machine, error) {
...
+	var err error
+	for i := 0; i < 5; i++ {
+		found, err := m.isNodeInAppliedSpec(obj)
+		if err == nil && !found {
+			break
+		}
+		if i == 4 {
+			return obj, fmt.Errorf("Machine [%s] still not deleted from cluster spec", obj.Name)
+		}
+		time.Sleep(1 * time.Second)
```
Comment (ibuildthecloud) [inline controller/machine/controller.go]: It is not a good idea block the goroutine. Rather just return the error and let the queue backoff.

File: controller/machine/controller.go [hunk 5]
```diff
@@ -300,3 +315,30 @@ func (m *Lifecycle) saveConfig(config *machineconfig.MachineConfig, machineDir s
...
+	}
+	if cluster.DeletionTimestamp != nil {
+		return false, nil
+	}
+	if cluster.Status.AppliedSpec.RancherKubernetesEngineConfig == nil {
+		return false, nil
+	}
+
+	for _, node := range cluster.Status.AppliedSpec.RancherKubernetesEngineConfig.Nodes {
+		machineName := strings.Split(node.MachineName, ":")
```
Comment (ibuildthecloud) [inline controller/machine/controller.go]: You should check the full string.  The format is `machine.Namespace + ":" + machine.Name`

**Pull request merged**